<a href="https://colab.research.google.com/github/asianli979/LinearAlgebraCode/blob/master/Copy_of_Assignment1_solutions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Jacob Chodoriwsky

import sympy as sp
import numpy as np
import math


# Question 1

In [ ]:
# Define the matrix as given. 
a = sp.symbols('a')
A = sp.Matrix([[2,3,-1,-2], [4*a,6*a,-2*a,-4*a], [4-4*a,6-6*a,2*a-2,4*a-4]])

A

Matrix([
[      2,       3,      -1,      -2],
[    4*a,     6*a,    -2*a,    -4*a],
[4 - 4*a, 6 - 6*a, 2*a - 2, 4*a - 4]])

In [ ]:
# Get RREF, there is only one nonzero row in RREF so the rank is 1. 
A.rref()

(Matrix([
 [1, 3/2, -1/2, -1],
 [0,   0,    0,  0],
 [0,   0,    0,  0]]), (0,))

# Question 2

# NOTE: 
There was a typo in answer. Based on the following analysis if you choose either of the following answers you received mark

- each bacon strip costs about $0.99

- each avocado costs about $1.19

<p>Four guys ordered four custom burgers at a restaurant that has specific prices per topping. The bun and patty cost $2.44. 

•	Mike’s burger cost $8.44 and had 2 avocados, 5 bacon strips, and 3 cheese slices. 

•	Gina’s burger cost $7.53 and had 3 avocados, 4 cheese slices, and 2 dill pickles. 

•	Tim’s burger cost $6.89 and had 2 bacon strips, one cheese slices, and 4 dill pickles. 

•	Lana’s burger cost $9.02 and had 4 avocados, 2 bacon strips, and a dill pickles. 

What is the per-topping cost of each topping above? Round to the nearest cent.  </p>


Let $a, b, c, d$ represent the cost per avocado, bacon strip, cheese slice, and dill pickle, respectively. Then: 

$$2a + 5b + 3c = 8.44$$
$$3a + 4c + 2d = 7.53$$
$$2b + c + 4d = 6.89$$
$$4a + 2b + d=9.02$$

In [ ]:
aug = sp.Matrix([[2,5,3,0,8.44], [3,0,4,2,7.53], [0,2,1,4,6.89], [4,2,0,1,9.02]])
aug

Matrix([
[2, 5, 3, 0, 8.44],
[3, 0, 4, 2, 7.53],
[0, 2, 1, 4, 6.89],
[4, 2, 0, 1, 9.02]])

In [ ]:
aug.rref()

(Matrix([
 [1, 0, 0, 0,  1.46652661064426],
 [0, 1, 0, 0,  0.98624649859944],
 [0, 0, 1, 0, 0.191904761904762],
 [0, 0, 0, 1,  1.18140056022409]]), (0, 1, 2, 3))

<p>From the RREF of the augmented matrix for the system, we see that: </p>
    
    - each avocado costs about $1.47 
    
    - each bacon strip costs about $0.99  (closest value to 0.99)
    
    - each cheese slice costs about $0.19
    
    - and each dill pickle costs about $1.18

If you subtract bun and patty cost $2.44 from the burger:

$$2a + 5b + 3c = 6$$
$$3a + 4c + 2d = 5.09$$
$$2b + c + 4d = 4.45$$
$$4a + 2b + d= 6.58$$

In [ ]:
aug = sp.Matrix([[2,5,3,0,6], [3,0,4,2,5.09], [0,2,1,4,4.45], [4,2,0,1,6.58]])
aug

Matrix([
[2, 5, 3, 0,    6],
[3, 0, 4, 2, 5.09],
[0, 2, 1, 4, 4.45],
[4, 2, 0, 1, 6.58]])

In [ ]:
aug.rref()

(Matrix([
 [1, 0, 0, 0,   1.10428571428571],
 [0, 1, 0, 0,  0.712857142857143],
 [0, 0, 1, 0, 0.0757142857142862],
 [0, 0, 0, 1,  0.737142857142857]]), (0, 1, 2, 3))

    - each avocado costs about $1.10  (closest value to 1.19)
    
    - each bacon strip costs about $0.71
    
    - each cheese slice costs about $0.08
    
    - and each dill pickle costs about $0.73

# Question 3

Suppose $A,B,C,$ and $D$ are all $2×2$ matrices. Find an expression for $B$ if $(A + B^T)C^{-1} = D$.

Right-multiply by $C^{-1}$:

$(A + B^T)C^{-1}C = DC$

$A + B^T = DC$

Subtract $A$: 

$A + B^T - A = DC - A$

$B^T = DC - A$

Transpose: 

$(B^T)^T = (DC - A)^T$

$B = (DC - A)^T$

# Question 4
Is the inverse of the given matrix a scalar multiple of $I_2$? Explain.

In [ ]:
x = sp.symbols('x')
M = sp.Matrix([[0, 1/x], [1/x, 0]])
M

Matrix([
[  0, 1/x],
[1/x,   0]])

In [ ]:
M.inv()

Matrix([
[0, x],
[x, 0]])

In [ ]:
M.inv()/x

Matrix([
[0, 1],
[1, 0]])

In [ ]:
I = sp.Matrix([[1,0], [0,1]])
I

Matrix([
[1, 0],
[0, 1]])

In [ ]:
M.inv()/x == I

False

<p># The given matrix is a scalar multiple of a non-identity matrix that looks a lot like I, but is not I. </p>

# Question 5

In [ ]:
A = np.ones((5,5))
A[0,3] = 2
A[1,1] = 2
A[3,0] = 2

A

array([[1., 1., 1., 2., 1.],
       [1., 2., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [2., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

In [ ]:
B = 3 - A.T
B

array([[2., 2., 2., 1., 2.],
       [2., 1., 2., 2., 2.],
       [2., 2., 2., 2., 2.],
       [1., 2., 2., 2., 2.],
       [2., 2., 2., 2., 2.]])

In [ ]:
b = [-2,0,5,1,3]
b

[-2, 0, 5, 1, 3]

In [ ]:
C = AB = np.dot(A,B)
C

array([[10., 11., 12., 11., 12.],
       [11., 10., 12., 11., 12.],
       [ 9.,  9., 10.,  9., 10.],
       [11., 11., 12., 10., 12.],
       [ 9.,  9., 10.,  9., 10.]])

In [ ]:
np.linalg.det(AB)

0.0

The determinat of A is Zero, and the answer does NOT exist for $x$. You can also calculate RREF and get to the same conclusion.

# Question 6

In [ ]:
r = sp.Matrix([2,5,x])
r

Matrix([
[2],
[5],
[x]])

$|r|=\sqrt{2^2 + 5^2 + x^2}$

$|r| = \sqrt{29 + x^2}$

$6 = \sqrt{29 + x^2}$

$36 = 29 + x^2$

$x^2 = 7$

$x = \pm \sqrt{7}$

# Question 7

Line 1: 

$x_1 = 1 + 5t$

$y_1 = -7 + 4t$

$z_1 = 2 - 3t$

Line 2: 

$x_2 = 2 - s$

$y_2 = -16 + 2s$

$z_2 = s$

Equate:

$1 + 5t = 2 - s$ simplifies to $s + 5t = 1$

$-7 + 4t = -1 + 2s$ simplifies to $-2s + 4t = 6$

$2 - 3t = s$ simplifies to $s + 3t = 2$

In [ ]:
[s,t] = sp.symbols('s,t')

M = sp.Matrix([[1,5,1], [2,-4,9], [1,3,2]])
M

Matrix([
[1,  5, 1],
[2, -4, 9],
[1,  3, 2]])

In [ ]:
M.rref()

(Matrix([
 [1, 0,  7/2],
 [0, 1, -1/2],
 [0, 0,    0]]), (0, 1))

We have a unique solution for $s,t$. The lines intersect where $s=7/2$ and $t=-1/2$. 

Substitute $s$ into Line 2 or $t$ into line 1 to get: 

$x = 1 + 5(-1/2)$

$x = -3/2$

$y = -7 + 4(-1/2)$

$y = -9$

$z = 2 - 3(-1/2)$

$z = 7/2$

The lines cross at the point $(-3/2, -9, 7/2)$. 